target/msm8974/init.c

```diff
@@ -149,15 +150,85 @@ uint32_t target_volume_down()
 		return pm8x41_resin_status();
 }
 
+#if WITH_LK2ND
+static struct lk2nd_keymap *keymap = NULL;
+
+static void target_init_keys(void)
+{
+	int i = 0;
+	struct pm8x41_gpio gpio = {
+		.direction = PM_GPIO_DIR_IN,
+		.function  = 0,
+		.vin_sel   = 2,
+	};
+	keymap = lk2nd_fdt_parse_early_keys();
+
+	while (keymap && keymap[i].key) {
+		if (keymap[i].gpio) {
+			gpio.pull = keymap[i].pull;
+			pm8x41_gpio_config(keymap[i].gpio, &gpio);
+		}
+		i++;
+	}
+
+	/* Wait for the gpio config to take effect - debounce time */
+	udelay(10000);
+}
+#endif
+
+static int target_key_pressed(int key)
+{
+	int i = 0;
+	uint8_t tmp = 0;
+#if WITH_LK2ND
+	while (keymap && keymap[i].key) {
+		if (keymap[i].key == key) {
+			if (!keymap[i].gpio)
+				return target_volume_down_old();
+			else {
+				pm8x41_gpio_get(keymap[i].gpio, &tmp);
+				if(tmp == keymap[i].active)
+					return 1;
+			}
+		}
+		i++;
+	}
+#endif
+	/* fallback to default handlers */
+	switch (key) {
+		case KEY_VOLUMEDOWN:
+			return target_volume_down_old();
+		case KEY_VOLUMEUP:
+			return target_volume_up_old();
+	}
+
+	return 0;
+}
+
+int target_volume_up() {
+    return target_key_pressed(KEY_VOLUMEUP);
+}
+
+int target_volume_down() {
+    return target_key_pressed(KEY_VOLUMEDOWN);
+}
```

Comment:
I wonder if we can (and if it's worth) de-duplicating this. It seems fairly similar for both 8916+8974.

Reply:
Something like `target_display_cont_splash_mdp5.c` maybe. Make a `target_keys_lk2nd.c`, define `target_volume_up/down` there and then guard the original definitions with `#if !WITH_LK2ND`. Would this work?

Reply:
Then how would I reuse those methods for, e.g. fallback?
Like this?
```
#if LK2ND
int foo_old()
#else
int foo()
#endif
{
...
```

Reply:
And then IMO 8916/8974 are pretty different: 8916 is gpio + resin and 8974 is pmic_gpio and (resin/resin_v2)

Reply:
Are the buttons even necessarily on PMIC on 8974? Sounds kind of like the typical unfortunate lk2nd assumption that makes one wish we had invested time in something more flexible than lk2nd (with proper device trees) :D

Reply:
I think fp2 needs pmic button looking at the mainline dt, and looking at a couple more + the fact that the target uses that makes me assume that *most* of 8974 use pmic gpio for the buttons (e.g.  klte uses 3 for home...)

Reply:
Maybe we should make this configurable though while we're at it? The PMIC GPIOs could also be used on 8916 theoretically and the other way around for 8974. The way to access SoC/PMIC GPIOs is pretty much the same for 8916/8974 I think?

Reply:
Hm, do you suggest something like 
```
#define PM_GPIO(x) (x | (1 << 16))
```
and check for the flag in the target handler?

Reply:
Well I don't know, perhaps we might as well define a dummy

```
/ {
    msmgpio: gpio {
        phandle = <0xMAGIC1>;
    };
    pmic_gpio: pmic-gpio {
        phandle = <0xMAGIC2>;
    };
}
```
and then properly do

```
something-gpios = <&msmgpio 42 GPIO_ACTIVE_HIGH>;
```

where the first uint32_t will end up as either 0xMAGIC1 or 0xMAGIC2.

Or we abandon lk2nd and do this properly with U-Boot.

Reply:
> Or we abandon lk2nd and do this properly with U-Boot.

You're going to start a revolution with statements like this :fire: